### PR TITLE
Add integration test for servers bound to localhost

### PR DESCRIPTION
**Pull request by adleong**
TestLocalhostServer creates an nginx deployment which listens on localhost
and a slow-cooker which attempts to send traffic to the nginx.  Since
slow-cooker should not be able to connect to nginx's localhost address,
these requests should fail.

This test currently asserts that these requests should succeed.  Once https://github.com/linkerd/linkerd2/issues/6495 is fixed, this test should be updated to assert that these requests should fail.

Signed-off-by: Alex Leong <alex@buoyant.io>

### CODE REVIEW

**Review by alpeb — COMMENTED** [review]
Assuming the proxy would return a 403 on policy failures, I tested this by changing the nginx config to have it return 403 instead, and `linkerd stat` still considers that a 100% success. And indeed I see metrics like `response_total{direction="outbound"... status_code="403" classification="success"} xxx` on the slow-cooker proxy. I understand the inbound proxy in nginx will probably flag these as failures, but how would the client do a proper classification?

File: test/integration/localhost/localhost_test.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+			}
+
+			if *stats[0].Rps == 0.0 {
+				return fmt.Errorf("expected non-zero RPS from slowcooker to nginx: %s", out)
+			}
+
+			// TODO: Update this test to expect these requests to fail once
+			// https://github.com/linkerd/linkerd2/issues/6495 is fixed.  Requests
+			// sent to a port which is only bound to localhost should fail.
+			if *stats[0].Success != 1.0 {
```
Comment (dadjeibaah) [inline test/integration/localhost/localhost_test.go:91]: Should we be also adding an additional check here that somehow shows that we didn't create any TCP connections between the proxy and the application on localhost? From what I gather, we are adding this proxy change which also add new metrics to show whether there were tcp accept errors. https://github.com/linkerd/linkerd2-proxy/pull/1161. Might be worth checking for that as well but not totally sure.

**Review by dadjeibaah — APPROVED** [review]
This looks great! I left a few questions but nothing major.

**Comment by adleong** [comment]
@alpeb good question.  I guess I'm still unclear on what the exact proxy behavior will be.  If it will return 403s in this case, we could add a response classifier in a service profile to classify the 403s as failures.

**Review by alpeb — APPROVED** [review]
LGTM, auth implementation will tell if adjustments will have to be made later :+1:

**Review by kleimkuhler — APPROVED** [review]
Looks good 👍

**Pull request merged**